File: Java/Salp.java

public class Salp extends Problem{
    private int[] x = new int[nVars];

    public Salp (){
		for ( int j = 0; j < nVars; j++ ) {
			x[j] = StdRandom.uniform(2);
		};
    };

    protected void leaderMove( Salp food, double c1 ){
        for ( int j = 0; j < nVars; j++ ){
            double c2 = Math.random();//necesario, definido en el Swarm no mejora el fitness 
            double c3 = Math.random() - 0.5;
            x[j] = toBinary( c3 >= 0 ? food.x[j] + c1*c2 : food.x[j] - c1*c2 );
        };
    };
    
    protected void followerMove( Salp previousFollower ){
        for ( int j = 0; j < nVars; j++ ){
            x[j] = toBinary( (0.5)* (x[j] + previousFollower.x[j]) );
            //x[j] = toBinary( Math.random() * (x[j] + previousFollower.x[j]) + Math.random() );
        };

    };

    protected boolean isFeasible() {
        return checkConstraint(x);
    };

    protected int computeFitness() {
		return computeFitness(x);
	};

    protected boolean isBetterThan (Salp g){
        return computeFitness() < g.computeFitness();
    };

    protected void copy(Object object){
        if ( object instanceof Salp ){
			System.arraycopy(((Salp) object).x, 0, this.x, 0, nVars);
        };
    };

	private String showSolution() {
		return java.util.Arrays.toString(x);
	}

	private float diff() {
		return computeFitness(x) - optimum();
	}

	private float rpd() {
		return diff() / optimum() * 100f;
	}

    private int toBinary(double x) {
		return StdRandom.uniform(0.5, 0.731059) <= (1 / (1 + Math.pow(Math.E, -x))) ? 1 : 0;
	};

    @Override
	public String toString() {
        return String.format("optimal value: %d, fitness: %d, diff: %.1f, rpd: %.2f%%, p: %s\n", optimum(),
            computeFitness(), diff(), rpd(), showSolution());
	}
    
};

File: Python/Salp.py
import math
import random
from Problem import Problem

class Salp( Problem ):

    def __init__( self ):
        super().__init__()
        self.x = [ random.randint(0, 1) for x in range(self.n_vars) ]
    
    def leader_move( self, food, c1 ):
        for index in range( 0, self.n_vars ):
            c2 = random.uniform(0, 1)
            c3 = random.uniform(0, 1) - 0.5
            self.x[index] = self.__to_binary( food.x[index] + c1*((1 - 0)*c2 + 0) if c3 >= 0 else food.x[index] - c1*((1 - 0)*c2 + 0) )

    def follower_move( self , previus_follower ):
        for index in range( 0, self.n_vars ):
            self.x[index] = self.__to_binary( 0.5*( self.x[index] + previus_follower.x[index] ) )
            #self.x[index] = self.__to_binary( random.random() * ( self.x[index] + previus_follower.x[index] ) + random.random() )
    
    def is_feasible( self ):
        return self.check_constraint(self.x)
    
    def compute_values_fitness( self ):
        return self.compute_fitness(self.x)
    
    def is_better_than( self, g ):
        return self.compute_values_fitness() < g.compute_values_fitness()
    
    def copy( self, object ):
        #if ...
        self.x = object.x.copy()
    
    def __to_binary( self, x: float ):
        return 1 if random.uniform(0.5, 0.731059) <= (1 / (1 + math.pow(math.e, -x))) else 0
